typed-function.js

```diff
@@ -655,28 +664,22 @@
      *                  or zero when both are equal
      */
     function compareParams (param1, param2) {
-      var c;
-
-      // compare having a rest parameter or not
-      c = param1.restParam - param2.restParam;
-      if (c !== 0) {
-        return c;
-      }
-
-      // compare having conversions or not
-      c = hasConversions(param1) - hasConversions(param2);
-      if (c !== 0) {
-        return c;
-      }
-
-      // compare the index of the types
-      c = getLowestTypeIndex(param1) - getLowestTypeIndex(param2);
-      if (c !== 0) {
-        return c;
+      // We compare a number of metrics on a param in turn:
+      const metrics = [
+        hasAny,
+        par => par.restParam,
+        hasConversions,
+        getLowestTypeIndex,
+        getLowestConversionIndex
+      ]
+
+      for (const metric of metrics) {
+        const c = metric(param1) - metric(param2)
+        if (c !== 0) return c
       }
 
-      // compare the index of any conversion
-      return getLowestConversionIndex(param1) - getLowestConversionIndex(param2);
+      // Don't have a basis for preference
```

Comment:
The only other way I could think of to break ties (if we wanted to minimize them) was to compare other type indices besides the lowest, but it didn't seem worth it because if I understand the rest of the code, it will only affect rest parameters; all the other params have been split by this point so that there's only a single type for each param.

In the increasingly common situation that Array.sort() is stable (that requirement is in the latest ECMA standard) it's fine to end up with no comparison between signatures, as then the original order will prevail. If you want me to enhance the code so that the signature sort will be stable even if Array.sort() is not, just let me know. It just didn't seem worth it on the first pass because it will be rare that non-identical signatures won't end up with some preference, and the JavaScript ecosystem is heading toward stable sort anyway.

Reply:
OK.